rust/automerge/src/marks.rs

```diff
@@ -21,6 +23,129 @@ pub struct Mark<'a> {
     pub(crate) data: Cow<'a, MarkData>,
 }
 
+impl<'a> Mark<'a> {
+    pub(crate) fn len(&self) -> usize {
+        self.end - self.start
+    }
+    pub(crate) fn into_mark_set(self) -> MarkSet {
+        let mut m = MarkSet::default();
+        let data = self.data.into_owned();
+        m.insert(data.name, data.value);
+        m
+    }
+}
+
+#[derive(Debug, Clone)]
+struct MarkAccItem {
+    index: usize,
+    len: usize,
+    value: ScalarValue,
+}
+
+#[derive(Debug, Clone, Default)]
+pub(crate) struct MarkAccumulator {
+    marks: BTreeMap<SmolStr, Vec<MarkAccItem>>,
+}
+
+impl MarkAccumulator {
+    pub(crate) fn into_iter(self) -> impl Iterator<Item = Mark<'static>> {
+        self.marks.into_iter().flat_map(|(name, items)| {
+            items.into_iter().map(move |i| {
+                Mark::new(name.to_string(), i.value.clone(), i.index, i.index + i.len)
+            })
+        })
+    }
+
+    pub(crate) fn into_iter_no_unmark(self) -> impl Iterator<Item = Mark<'static>> {
+        self.marks.into_iter().flat_map(|(name, items)| {
+            items
+                .into_iter()
+                .filter(|i| !i.value.is_null())
+                .map(move |i| {
+                    Mark::new(name.to_string(), i.value.clone(), i.index, i.index + i.len)
+                })
+        })
+    }
+
+    pub(crate) fn add(&mut self, index: usize, len: usize, marks: &MarkSet) {
+        for (name, value) in marks.marks.iter() {
+            let entry = self.marks.entry(name.clone()).or_default();
+            if let Some(mut last) = entry.last_mut() {
+                if &last.value == value && last.index + last.len == index {
+                    last.len += len;
+                    continue;
+                }
+            }
+            entry.push(MarkAccItem {
+                index,
+                len,
+                value: value.clone(),
+            })
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Default)]
+pub struct MarkSet {
+    marks: Arc<BTreeMap<SmolStr, ScalarValue>>,
+}
+
+#[derive(Debug)]
+pub struct MarkSetIterator<'a> {
+    iter: std::collections::btree_map::Iter<'a, SmolStr, ScalarValue>,
+}
+
+impl<'a> Iterator for MarkSetIterator<'a> {
+    type Item = (&'a str, &'a ScalarValue);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.iter.next().map(|(name, value)| (name.as_str(), value))
+    }
+}
+
+impl MarkSet {
+    pub fn iter(&self) -> MarkSetIterator<'_> {
```

Comment:
Can we just return `impl Iterator<..>` here and avoid the manually implemented `MarkSetIterator`